packages/shared/hooks/useAsync.ts

```diff
@@ -58,47 +56,106 @@ import { useCallback, useState } from 'react';
  *       return <UserAvatar url={fetchUserProfileAttempt.data.avatarUrl} />;
  *    }
  * }
+ *
+ * @example Aborting and useEffect cleanup function.
+ * If the first argument passed to the run function is an abort signal and the signal gets aborted,
+ * useAsync will ignore the return value of the promise from the callback.
+ *
+ * This lets you write a cleanup function for useEffect:
+ *
+ * const [attempt, run] = useAsync((message: string) => doAsyncStuff(message));
+ *
+ * useEffect(() => {
+ *   const abortController = new AbortController();
+ *   run(abortController.signal, 'hello');
+ *
+ *   return () => {
+ *     abortController.abort();
+ *   };
+ * });
  */
-export function useAsync<R, T extends Function>(cb?: AsyncCb<R, T>) {
-  const [state, setState] = useState<Attempt<R>>(makeEmptyAttempt);
+export function useAsync<Callback extends (...args: unknown[]) => unknown>(
```

Comment:
Hmm, I think the solution with `isMounted` doesn't handle situations where there's no unmounting but we simply want to ignore the result of the fetch because, let's say, params for the fetch query have changed.

I forgot to include `run` in `useEffect` deps in the example, but also imagine that the `message` passed to `run` could change when the user changes an input field. `isMounted` wouldn't handle that situation, would it?

Reply:
No, isMounted only handles unmount cases. 
The solution to the problem you referring to is 
```
if (promise !== asyncTask.current) return // [...] 
```
check.
This will ensure that only the latest promise is handled by then/catch metchods

Reply:
Ah, I completely missed that while looking at your solution.

Alright, this seems to be even better! I'm trying to think of a situation where we wouldn't want to return early on unmount or the promise being changed but nothing comes to my mind right now.

Another thing is that this solution would change the default behavior of `useAsync` but I think that's for the better - it'll automatically clean up those calls in flight. Places which depend solely on `attempt` won't see a difference and places which depend on the resolved value of the promise should handle errors anyway.

Thanks a lot!

Reply:
And if a certain place doesn't need to ignore stale request, it'd be straightforward to add an option to `useAsync` which simply ignores those checks and does allow stale requests.

Reply:
> it'd be straightforward to add an option to useAsync which simply ignores those checks and does allow stale requests.

In this case, we can also return current promise instead:
```
if (promise !== asyncTask.current) return asyncTask.current.then(...)
```